### PR TITLE
Changes for R14

### PR DESCRIPTION
**Pull request by tsuwandy**
Issue #1252 
- Incremental update for bf orchestrator:create  (if existing blu file exists)
- Added bf orchestrator:remove command
- Added download progress info for Orchestrator base model download onProgress callback

### CODE REVIEW

File: packages/orchestratorlib/src/basemodel.ts [hunk 1]
```diff
@@ -100,13 +100,25 @@ export class OrchestratorBaseModel {
...
 
-  public static async downloadModel(modelUrl: string, modelZipPath: string): Promise<(resolve: any) => void> {
+  public static async downloadModel(modelUrl: string, modelZipPath: string, onProgress: any = OrchestratorBaseModel.defaultHandler): Promise<(resolve: any) => void> {
     const response: AxiosResponse<Stream> = await axios({
       method: 'GET',
       url: modelUrl,
       responseType: 'stream',
     });
+    const totalLength: number = response.headers['content-length'];
+    onProgress(`Total to download: ${totalLength} bytes...`);
```
Comment (daveta) [inline packages/orchestratorlib/src/basemodel.ts:110]: Neat.

**Review by daveta — APPROVED** [review]

File: packages/orchestrator/src/commands/orchestrator/add.ts [hunk 2]
```diff
@@ -57,33 +57,32 @@ export default class OrchestratorAdd extends Command {
         fullEmbeddings = true;
       }
 
+      const settings: OrchestratorSettings = OrchestratorSettings.getCurrent();
```
Comment (joshgummersall) [inline packages/orchestrator/src/commands/orchestrator/add.ts:60]: Is the return type of `getCurrent` already `OrchestratorSettings`? Generally, I prefer to omit type annotations that are automatically inferred by the compiler. You should be able to write:
```typescript
const settings = OrchestratorSettings.getCurrent();
```

Comment (joshgummersall) [reply packages/orchestrator/src/commands/orchestrator/add.ts:60]: I noticed this in a lot of places so I won't comment on all of them, it's definitely something that could get cleaned up if there's time.

File: packages/orchestrator/src/commands/orchestrator/create.ts [hunk 3]
```diff
@@ -55,45 +56,45 @@ export default class OrchestratorCreate extends Command {
 
       if (DataSourceHelper.isDispatchInput(input)) {
         const dispatchJson: any = JSON.parse(OrchestratorHelper.readFile(input));
-        DataSourceHelper.convertDispatchInputs(dispatchJson, OrchestratorSettings.DataSources);
+        DataSourceHelper.convertDispatchInputs(dispatchJson, settings.DataSources);
         refresh = true;
       }
 
-      const hasDataSources: boolean = OrchestratorSettings.DataSources && OrchestratorSettings.DataSources.inputs.length > 0;
+      const hasDataSources: boolean = settings.DataSources && settings.DataSources.inputs.length > 0;
```
Comment (joshgummersall) [inline packages/orchestrator/src/commands/orchestrator/create.ts:63]: You can write this as `const hasDataSources = settings.DataSources?.inputs?.length > 0 ?? false;`

File: packages/orchestrator/src/commands/orchestrator/create.ts [hunk 4]
```diff
@@ -55,45 +56,45 @@ export default class OrchestratorCreate extends Command {
...
+      settings.persist();
     } catch (error) {
       throw (new CLIError(error));
     }
   }
 
-  private async refreshLuisQnAInputs(dataSources: OrchestratorDataSourceSettings): Promise<void> {
-    if (!dataSources) {
+  private async refreshLuisQnAInputs(settings: OrchestratorSettings): Promise<void> {
+    if (!settings || !settings.DataSources) {
```
Comment (joshgummersall) [inline packages/orchestrator/src/commands/orchestrator/create.ts:89]: `if (!settings?.Datasources) {}`

File: packages/orchestrator/src/commands/orchestrator/remove.ts [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import * as path from 'path';
+import {Command, CLIError, flags} from '@microsoft/bf-cli-command';
+import {Orchestrator, OrchestratorDataSource, OrchestratorSettings} from '@microsoft/bf-orchestrator';
+
+export default class OrchestratorRemove extends Command {
+  static description: string = 'Remove examples from LUIS app(s), QnaMaker kb(s) or .lu/.qna/.json files from Orchestrator snapshot file.';
```
Comment (joshgummersall) [inline packages/orchestrator/src/commands/orchestrator/remove.ts:11]: `: string` is inferred.

File: packages/orchestrator/src/commands/orchestrator/remove.ts [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+
+  static examples: Array<string> = [`
```
Comment (joshgummersall) [inline packages/orchestrator/src/commands/orchestrator/remove.ts:13]: `: Array<string>` is inferred. Is this meant to be a list of separate examples, or a single item list of one multiline example?

File: packages/orchestrator/src/commands/orchestrator/remove.ts [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+        } else {
+          this.log(`Removed ${dataSource.Type} source with id ${dataSource.Id}`);
+        }
+      } else {
+        this.log('Data source not found');
+      }
+
+      settings.persist();
+    } catch (error) {
+      throw (new CLIError(error));
```
Comment (joshgummersall) [inline packages/orchestrator/src/commands/orchestrator/remove.ts:55]: Parens around `new CLIError` are unnecessary.

File: packages/orchestratorlib/src/build.ts [hunk 8]
```diff
@@ -279,6 +275,12 @@ export class OrchestratorBuild {
     });
   }
 
+  // Synchronize an active LabelResolver instance with an LU file.
+  static async syncLabelResolver(labelResolver: LabelResolver, luContent: string) {
+    const target: ITextUtteranceLabelMapDataStructure = await OrchestratorBuild.getExamplesLU(luContent);
+    OrchestratorBuild.syncLabelResolverEx(labelResolver, target);
```
Comment (joshgummersall) [inline packages/orchestratorlib/src/build.ts:281]: You should either `await` this function, too, or `return OrchestratorBuild.syncLabelResolverEx(...)` so the promise is propagated.

File: packages/orchestratorlib/src/create.ts [hunk 9]
```diff
@@ -11,6 +11,7 @@ import {ITextUtteranceLabelMapDataStructure} from '@microsoft/bf-dispatcher';
 // import {Label} from '@microsoft/bf-dispatcher';
 
 import {LabelResolver} from './labelresolver';
+import {OrchestratorBuild} from '.';
```
Comment (joshgummersall) [inline packages/orchestratorlib/src/create.ts:14]: Generally, I think it's better to use the file path here rather than implicitly referencing `index.ts`.

File: packages/orchestratorlib/src/create.ts [hunk 10]
```diff
@@ -45,42 +46,16 @@ export class OrchestratorCreate {
...
-     *    } else {
-     *      Utility.debuggingLog('');
-     *    }
-     *    for (const label of labels) {
-     *      Utility.debuggingLog(`label=${label.name}`);
-     *    }
-     *  }
-     */
+
+    OrchestratorBuild.syncLabelResolverEx(labelResolver, processedUtteranceLabelsMap);
```
Comment (joshgummersall) [inline packages/orchestratorlib/src/create.ts:58]: `await`?

File: packages/orchestratorlib/src/build.ts [hunk 11]
```diff
@@ -154,14 +151,13 @@ export class OrchestratorBuild {
     return luIntentsEntitiesUtterances;
   }
 
-  // Synchronize an active LabelResolver instance with with an LU file.
-  static async syncLabelResolver(labelResolver: LabelResolver, luContent: string) {
+  // Synchronize an active LabelResolver instance with ITextUtteranceLabelMapDataStructure.
+  static async syncLabelResolverEx(labelResolver: LabelResolver, target: ITextUtteranceLabelMapDataStructure) {
```
Comment (joshgummersall) [inline packages/orchestratorlib/src/build.ts:155]: Also, add `: Promise<void>` as the return type here.

File: packages/orchestratorlib/src/build.ts [hunk 12]
```diff
@@ -279,6 +275,12 @@ export class OrchestratorBuild {
     });
   }
 
+  // Synchronize an active LabelResolver instance with an LU file.
+  static async syncLabelResolver(labelResolver: LabelResolver, luContent: string) {
```
Comment (joshgummersall) [inline packages/orchestratorlib/src/build.ts:279]: Add `: Promise<void>` as the return type.

File: packages/orchestratorlib/src/settings.ts [hunk 13]
```diff
@@ -124,23 +161,33 @@ export class OrchestratorSettings {
...
   }
 
-  public static addUpdateDataSource(data: OrchestratorDataSource) {
+  public addUpdateDataSource(data: OrchestratorDataSource) {
     this.DataSources.inputs.push(data);
   }
 
-  public static ModelPath: string;
+  public static getCurrent(): OrchestratorSettings {
+    if (!OrchestratorSettings.Current) {
```
Comment (joshgummersall) [inline packages/orchestratorlib/src/settings.ts:173]: `OrchestratorSettings.Current ??= new OrchestratorSettings();`

File: packages/orchestratorlib/src/settings.ts [hunk 14]
```diff
@@ -152,59 +199,57 @@ export class OrchestratorSettings {
...
-  public static init(
+  public init(
     settingsDir: string,
     baseModelPath: string,
     entityBaseModelPath: string,
     snapshotPath: string,
     hierarchical: boolean = false,
-    hasDataSources: boolean = false)  {
+    hasDataSources: boolean = false,
+    settingsFileName: string = OrchestratorSettings.OrchestratorSettingsFileName)  {
```
Comment (joshgummersall) [inline packages/orchestratorlib/src/settings.ts:209]: Add `: void` return type.

File: packages/orchestratorlib/tsconfig.json [hunk 15]
```diff
@@ -9,6 +9,7 @@
     "module": "commonjs",
     "outDir": "./lib",
     "pretty": true,
+    "strictNullChecks": false,
```
Comment (joshgummersall) [inline packages/orchestratorlib/tsconfig.json:12]: `strictNullChecks` is a wonderful feature, why is this necessary?

**Review by taicchoumsft — APPROVED** [review]
:shipit:

File: packages/orchestrator/src/commands/orchestrator/add.ts (hunk 2 above)
Comment (tsuwandy) [reply packages/orchestrator/src/commands/orchestrator/add.ts:60]: yeah, it's annoying but we use eslint and it complains if I don't have the type in there... :-(

File: packages/orchestrator/src/commands/orchestrator/remove.ts (hunk 5 above)
Comment (tsuwandy) [reply packages/orchestrator/src/commands/orchestrator/remove.ts:11]: unfortunately, eslint is forcing this. :-(

File: packages/orchestrator/src/commands/orchestrator/remove.ts (hunk 6 above)
Comment (tsuwandy) [reply packages/orchestrator/src/commands/orchestrator/remove.ts:13]: same deal with the type annotation, thanks to the annoying eslint. :-)  

image

File: packages/orchestratorlib/src/build.ts (hunk 12 above)
Comment (tsuwandy) [reply packages/orchestratorlib/src/build.ts:279]: fixed!

File: packages/orchestratorlib/src/settings.ts (hunk 13 above)
Comment (tsuwandy) [reply packages/orchestratorlib/src/settings.ts:173]: ok :-)

File: packages/orchestratorlib/tsconfig.json (hunk 15 above)
Comment (tsuwandy) [reply packages/orchestratorlib/tsconfig.json:12]: ah, i wanted https://github.com/microsoft/botframework-cli/blob/8cae60197f53eab4dc93b4f5d63001356c811755/packages/orchestratorlib/src/settings.ts#L113 to return null and the caller to check for null.  Wasn't possible without it.  Thoughts on alternative?

File: packages/orchestratorlib/src/settings.ts (hunk 13 above)
Comment (tsuwandy) [reply packages/orchestratorlib/src/settings.ts:173]: uh, unfortunately eslint doesn't like OrchestratorSettings.Current ??= new OrchestratorSettings();

D:\src\botframework-cli\packages\orchestratorlib>npm run posttest

> @microsoft/bf-orchestrator@1.0.0 posttest D:\src\botframework-cli\packages\orchestratorlib
> eslint . --ext .ts --config .eslintrc

TypeError: Cannot read property 'range' of null
Occurred while linting D:\src\botframework-cli\packages\orchestratorlib\src\settings.ts:173
    at SourceCode.getTokenBefore (D:\src\botframework-cli\common\temp\node_modules\.pnpm\registry.npmjs.org\eslint\5.16.0\node_

reverting change.. :-(

File: packages/orchestratorlib/tsconfig.json (hunk 15 above)
Comment (tsuwandy) [reply packages/orchestratorlib/tsconfig.json:12]: ok, will use union type instead and remove "strictNullChecks": false,

File: packages/orchestrator/src/commands/orchestrator/remove.ts (hunk 7 above)
Comment (munozemilio) [reply packages/orchestrator/src/commands/orchestrator/remove.ts:55]: It would be nice to filter expected errors from Orchestrator otherwise telemetry will never catch unexpected issues

File: packages/orchestratorlib/src/settings.ts [hunk 16]
```diff
@@ -152,59 +198,57 @@ export class OrchestratorSettings {
...
+
+      this.SnapshotPath = OrchestratorSettings.ensureSnapshotPath(snapshotPath, settingsDir, settings);
+      this.ModelPath = OrchestratorSettings.ensureBaseModelPath(baseModelPath, settings);
       entityBaseModelPath = OrchestratorSettings.ensureEntityBaseModelPath(entityBaseModelPath, settings);
       if (!Utility.isEmptyString(entityBaseModelPath)) {
-        OrchestratorSettings.EntityModelPath = entityBaseModelPath;
+        this.EntityModelPath = entityBaseModelPath;
       }
-    }  catch {}
+    } catch {}
```
Comment (munozemilio) [inline packages/orchestratorlib/src/settings.ts:232]: Whats the purpose of an empty catch?

File: packages/orchestrator/src/commands/orchestrator/basemodel/get.ts [hunk 17]
```diff
@@ -64,10 +66,15 @@ export default class OrchestratorBaseModelGet extends Command {
...
             modelInfo.modelFolder,
             modelInfo.versionId,
-            (message: any) => {
-              if (flags.verbose) {
+            (message: any, pct: number) => {
+              if (message) {
                 this.log(message);
               }
+              if (flags.verbose) {
+                if (pct) {
```
Comment (munozemilio) [inline packages/orchestrator/src/commands/orchestrator/basemodel/get.ts:74]: Nit
```
if (flags.verbose
    && pct) { 
    ....
}
```
This could reduce indentation a bit.

**Review by munozemilio — APPROVED** [review]
I left some small comments.

File: packages/orchestrator/src/commands/orchestrator/basemodel/get.ts (hunk 17 above)
Comment (tsuwandy) [reply packages/orchestrator/src/commands/orchestrator/basemodel/get.ts:74]: makes sense! :-)
